### ISSUE TITLE
regexpSimplify gives an incorrect simplification for `(foo|)`.

### ISSUE THREAD

**Issue by csilvers** [go-critic/go-critic#1418]
For this file:
```
package test
import "regexp"
var _ = regexp.MustCompile(`(a|)`)
```

gocritic gives this output:
```
% go run github.com/go-critic/go-critic/cmd/gocritic@latest check --enable '#style' /tmp/test.go
/tmp/test.go:5:28: regexpSimplify: can re-write `(a|)` as `(|?)`
```

The simplification presumably was meant to be `(a)?`.

Another example is the string `(?:\.(\d+)|)`, which gocritic wants to simplify to `(?:|?)`.  Which may give some idea about what it's matching incorrectly here.

**Comment by olivergondza** [comment]
Found a specimen of the same: "can re-write `|[0-9a-zA-Z]{16}` as `|?` (gocritic)"

**Comment by nieomylnieja** [comment]
I reproduced this with golangci-lint v2.12.2, which bundles go-critic v0.14.3,
on Go 1.26.7 (linux/amd64).

Minimal reproduction:

`go.mod`:

```go
module example.com/regexp-repro

go 1.25.13
```

`main.go`:

```go
package main

import "regexp"

var _ = regexp.MustCompile(`^(?:|\d+)$`)

func main() {}
```

`.golangci.yml`:

```yaml
version: "2"
linters:
  default: none
  enable:
    - gocritic
  settings:
    gocritic:
      enabled-checks:
        - regexpSimplify
```

Run `golangci-lint run`. It reports:

```text
main.go:5:28: regexpSimplify: can re-write `^(?:|\d+)$` as `^(?:|?)$` (gocritic)
```

The original program succeeds with `go run .`.
Replacing the regex with the suggested value causes a panic during package
initialization:

```text
panic: regexp: Compile(`^(?:|?)$`): error parsing regexp: missing argument to repetition operator: `?`
```

I also traced the failure in v0.14.3:

- [`concatLiteral` and `factorPrefixSuffix`][factoring] mishandle the alternatives.
  The empty branch yields `"|"`, while the nonliteral `\d+` branch yields `""`.
  The factoring code treats both values as literal text and emits `|?`.
- [The second pass][passes] fails to parse that candidate.
  The loop stops but retains and reports the invalid first candidate.

For this specific pattern, `^\d*$` is an equivalent workaround and produces
no warning with the same linter.

[factoring]: https://github.com/go-critic/go-critic/blob/9aea378c4dccd6f4394196ad8f0873b3e84678c8/checkers/regexpSimplify_checker.go#L368-L419
[passes]: https://github.com/go-critic/go-critic/blob/9aea378c4dccd6f4394196ad8f0873b3e84678c8/checkers/regexpSimplify_checker.go#L72-L115
